msticpy/data/uploaders/loganalytics_uploader.py

```diff
@@ -0,0 +1,238 @@
+from .uploader_base import UploaderBase
+from ..._version import VERSION
+
+# Credits
+# https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api#python-3-sample
+# https://github.com/Cyb3rWard0g/azure-loganalytics-api-clients/blob/master/python/ala-python-data-producer.py
+
+
+__version__ = VERSION
+__author__ = "Pete Bryan"
+
+
+class LAUploader(UploaderBase):
+    """Uploader class for LogAnalytics."""
+
+    def __init__(self, workspace: str, workspace_secret: str, **kwargs):
+        """Initialize a LogAnalytics Uploader instance."""
+        super().__init__()
+        self._kwargs = kwargs
+        self.workspace = workspace
+        self.workspace_secret = workspace_secret
+        self._debug = kwargs.get("debug", False)
+        self.ops_loc = kwargs.get("opsinsight_loc", ".ods.opinsights.azure.com")
+
+    def _build_signature(
+        self,
+        date: str,
+        content_length: int,
+        method: str,
+        content_type: str,
+        resource: str,
+    ) -> str:
+        """
+        Build authentication authentication string to pass to LA API.
+
+        Parameters
+        ----------
+        date : str
+            datetime of authencation session.
+        content_length : int
+            lenght of content to be passed to the api.
+        method : str
+            HTTP method being used.
+        content_type : str
+            Type of content being passed to the API.
+        resource : str
+            The API endpoint being targetted.
+
+        Returns
+        -------
+        str
+            The encoded authorization string.
+
+        """
+        x_headers = "x-ms-date:" + date
+        string_to_hash = (
+            method
+            + "\n"
+            + str(content_length)
+            + "\n"
+            + content_type
+            + "\n"
+            + x_headers
+            + "\n"
+            + resource
+        )
+        bytes_to_hash = bytes(string_to_hash, encoding="utf-8")
+        decoded_key = base64.b64decode(self.workspace_secret)
+        encoded_hash = base64.b64encode(
+            hmac.new(decoded_key, bytes_to_hash, digestmod=hashlib.sha256).digest()
+        ).decode()
+        authorization = f"SharedKey {self.workspace}:{encoded_hash}"
+        return authorization
+
+    def _post_data(self, body: str, table_name: str):
+        """
+        Write data to Log Analytics Workspace.
+
+        Parameters
+        ----------
+        body : str
+            The JSON formatted data to write to Log Analytics.
+        table_name : str
+            The name of the custom table to write the data to.
+
+        Raises
+        ------
+        MsticpyConnectionError
+            Raised when response code indicates failure.
+
+        """
+        table_name = re.sub("[^A-Za-z0-9_]+", "", table_name)
+
+        resource = "/api/logs"
+        content_type = "application/json"
+        rfc1123date = datetime.datetime.utcnow().strftime("%a, %d %b %Y %H:%M:%S GMT")
```

Comment:
Does it really use "GMT"?

Reply:
Yep. The API 403s if you pass it anything else...